std/bigint.d

```diff
@@ -1025,6 +1025,212 @@ public:
+
+        Returns: a `ubyte[]` array that represents the native-endian
+            representation of that number.
+    */
+    ubyte[] toBytes() const pure @system @property nothrow
+    in
+    {
+        assert(this.uintLength > 0u, "A BigInt contained no underlying uints!");
+    }
+    out (ret)
+    {
+        assert(ret.length > 0u, "BigInt.toBytes returned zero bytes!");
+    }
+    do
+    {
+        ubyte[] ret = this.data.toBytes;
+        if (this.sign)
+        {
+            foreach (ref u; ret)
+            {
+                u = cast(ubyte) (~ cast(int) u);
+            }
+            BigInt.incrementBytes(ret);
+        }
+
+        version (BigEndian)
+        {
+            if (this.sign)
+            {
+                if (!(ret[$-1] & 0x80u)) ret = (0xFFu ~ ret);
+            }
+            else
+            {
+                if (ret[$-1] & 0x80u) ret = (0x00u ~ ret);
+            }
+        }
+        else version (LittleEndian)
+        {
+            if (this.sign)
+            {
+                if (!(ret[$-1] & 0x80u)) ret ~= 0xFFu;
+            }
+            else
+            {
+                if (ret[$-1] & 0x80u) ret ~= 0x00u;
+            }
+        }
+        else static assert(0, "Undetermined endianness! Cannot compile!");
+        return ret;
+    }
+
+    @system
+    unittest
+    {
+        BigInt b1 = BigInt("18446744073709551619"); // ulong.max + 4
+        BigInt b2 = BigInt(uint.max);
+        BigInt b3 = BigInt(0);
+        BigInt b4 = BigInt(3);
+        BigInt b5 = BigInt(-3);
+
+        version (LittleEndian)
+        {
+            assert(b1.toBytes == [
+                0x03u, 0x00u, 0x00u, 0x00u,
+                0x00u, 0x00u, 0x00u, 0x00u,
+                0x01u
+            ]);
+
+            assert(b2.toBytes == [
+                0xFFu, 0xFFu, 0xFFu, 0xFFu,
+                0x00u
+            ]);
+
+            assert(b3.toBytes == [
+                0x00u
+            ]);
+
+            assert(b4.toBytes == [
+                0x03u
+            ]);
+
+            assert(b5.toBytes == [
+                0xFDu
+            ]);
+        }
+        else version (BigEndian)
+        {
+            assert(b1.toBytes == [
+                0x01u, 0x00u, 0x00u, 0x00u,
+                0x00u, 0x00u, 0x00u, 0x00u,
+                0x03u
+            ]);
+
+            assert(b2.toBytes == [
+                0x00u, 0xFFu, 0xFFu, 0xFFu,
+                0xFFu
+            ]);
+
+            assert(b3.toBytes == [
+                0x00u
+            ]);
+
+            assert(b4.toBytes == [
+                0x03u
+            ]);
+
+            assert(b5.toBytes == [
+                0xFDu
+            ]);
+        }
+    }
+
+    /**
+        Override for the `cast` operator, so the `BigInt` can be cast to a
+        `ubyte[]`.
+
+        Returns: a `ubyte[]` array, representing the native-endian
+            representation of that number.
+    */
+    T opCast(T : ubyte[])() pure nothrow @system const
```

Comment:
I am a bit wary of an `opCast` that allocates. Can someone else comment on whether that's normal for Phobos?

Reply:
Searching for `opCast` in Phobos yields many results, but none of them seem to be allocating.

Reply:
I am a little confused. What is the alternative? This `opCast()` override just calls the `toBytes()` accessor.

Reply:
The alternative is to not have the `opCast` at all, and force the programmer to call `toBytes()` to get the byte representation. The advantage of this is the allocation is more explicit, while `cast(byte[])myBigint` may look like it's not doing anything scary - after all, for some possible implementations, it would simply return a different view of the same data.

Reply:
Ah. I've never actually thought about that, but that makes sense! I will remove the `opCast` override then.